pkg/registry/apis/query/query.go

```diff
@@ -217,17 +262,154 @@ func (b *QueryAPIBuilder) handleQuerySingleDatasource(ctx context.Context, req d
 		}
 	}
 
-	// Create a response object with the error when missing (happens for client errors like 404)
-	if rsp == nil && err != nil {
-		rsp = &backend.QueryDataResponse{Responses: make(backend.Responses)}
-		for _, q := range req.Request.Queries {
-			rsp.Responses[q.RefID] = backend.DataResponse{
-				Status: backend.Status(code),
-				Error:  err,
+	return code, rsp, elapsedTime, err
+}
+
+// peformComparison determines if a comparison check should be made
+func (b *QueryAPIBuilder) peformComparison(req datasourceRequest) bool {
+	// not in passive mode
+	if b.passiveModeClient == nil {
+		return false
+	}
+
+	// do not compare relative requests - they will always be different since they
+	// are served at different times
+	_, err := strconv.ParseInt(req.Request.TimeRange.From, 10, 64)
+	_, err2 := strconv.ParseInt(req.Request.TimeRange.To, 10, 64)
+	if (err != nil) && (err2 != nil) {
+		b.log.Debug("relative query - not running comparison")
+		return false
+	}
+
+	// Only compare Loki & Prometheus requests. We do not want to compare all requests
+	// as some datasources have costs attached to queries.
+	return req.PluginId == datasources.DS_PROMETHEUS || req.PluginId == datasources.DS_LOKI
```

Comment:
another nit: It's probably faster to check this so I'd check the pluginId before parsing the time range

Reply:
Good idea - updated!